netket/logging.py

```diff
@@ -0,0 +1,115 @@
+        are `[o]verwrite` (default), `[a]ppend`, `[n]ewfile` (creates a new file with an incremental
+        name) or `[f]ail`.
+    """
+
+    def __init__(
+        self, output_prefix, save_params_every=50, write_every=50, mode="overwrite"
+    ):
+        # Shorthands for mode
+        if mode == "o":
+            mode = "overwrite"
+        elif mode == "a":
+            mode = "append"
+        elif mode == "n":
+            mode = "newfile"
+        elif mode == "f":
+            mode = "fail"
+
+        if not (
+            (mode == "overwrite")
+            or (mode == "append")
+            or (mode == "newfile")
+            or (mode == "fail")
+        ):
+            raise ValueError(
+                "Mode not recognized: should be one of `[o]verwrite`, `[a]ppend`, `[n]ewfile or `[f]ail`."
+            )
+
+        file_exists = _exists_json(output_prefix)
+
+        starting_json_content = {"Output": []}
+
+        if file_exists and mode == "newfile":
+            i = 1
+            # Generate a new output_prefix with an incremental counter
+            while i < 1001:
+                _prefix = output_prefix + "_{}".format(i)
+                if not _exists_json(_prefix):
+                    output_prefix = _prefix
+                    break
+                i = i + 1
+
+            # give up after we tried too many files to avoid an infinite loop
+            if i is 1000:
+                raise IOError(
+                    "More than 1000 files exist with that prefix. Please change ouput_prefix manually."
+                )
+        elif file_exists and mode == "append":
+            # if there is only the .wf file but not the json one, raise an error
+            if not _path.exists(output_prefix + ".log"):
+                raise ValueError(
+                    "History file does not exists, but wavefunction file does. Please change `output_prefix or set mode=`overwrite`."
+                )
+
+            starting_json_content = _json.load(open(output_prefix + ".log"))
+
+        elif file_exists and mode == "fail":
+            raise ValueError(
+                "Output file already exists. Either delete it manually or change `output_prefix`."
+            )
+
+        self._json_out = starting_json_content
+        self._prefix = output_prefix
+        self._write_every = write_every
+        self._save_params_every = save_params_every
+        self._old_step = 0
+
+    def __call__(self, step, item, machine):
+        item["Iteration"] = step
+
+        self._json_out["Output"].append(item)
+
+        if step % self._write_every == 0 or step == self._old_step - 1:
+            self._flush_log()
+        if step % self._save_params_every == 0 or step == self._old_step - 1:
+            self._flush_params(machine)
+
+        self._old_step = step
+
+    def _flush_log(self):
+        with open(self._prefix + ".log", "w") as outfile:
+            _json.dump(self._json_out, outfile)
```

Comment:
Shouldn't this also only be done on MPI rank 0?

Reply:
My idea was that loggers are MPI-unaware, but the drivers knowingly only log/write/flush on rank 0. This simplifies implementing new loggers.

It's why I set `logger=None` in `abstractDriver.run`.

But if you think the opposite is more appropriate...

Reply:
Ah, I see. But `_flush_params` does check the MPI rank right now.

I think generally I agree with having an MPI-unaware logger.

Reply:
Ops, copy-paste leftover. Will remove that check asap